design/src/aws_config.md

```diff
@@ -0,0 +1,161 @@
+# AWS Configuration RFC
+
+> Status: RFC. For an ordered list of proposed changes see: [Proposed changes](#changes-checklist).
+
+An AWS SDK loads configuration from multiple locations. Some of these locations can be loaded synchronously. Some are
+async. Others may actually use AWS services such as STS or SSO.
+
+This document proposes an overhaul to the Config design to facilitate three things:
+
+1. Future-proof: It should be easy to add additional sources of region and credentials, sync and async, from many
+   sources, including code generated services.
+2. Ergonomic: There should be one obvious way to create an AWS service client. Customers should be able to easily
+   customize the client to make common changes. It should lead customers into the pit of success in terms of sharing
+   things that are expensive to create.
+3. Shareable: A config object should be usable to configure multiple AWS services.
+
+## New customer experience
+
+With the new shared config, customers must always take `aws-config` as a dependency if they want access to the automatic
+configuration loading.
+
+In Cargo.toml:
+
+```toml
+[dependencies]
+aws-sdk-dynamo = "0.1"
+aws-config = "0.5"
+```
+
+When creating a client, customers will create a `SharedConfig` struct, then use it to construct the service client.
+The `SharedConfig` struct is defined centrally in `aws-types`. (See [stability](#stability-and-versioning))
+
+```rust
+// pub use of aws_types::SharedConfig
+use aws_config::SharedConfig;
+
+async fn main() {
+    // loading a shared config may make network calls, etc.
+    // with some custom configuration:
+    let config = aws_config::config_loader()
+            .region(Region::new("us-east-1"))
+            .sleep(async_std_sleep)
+            .load().await;
+
+    // with no custom configuration:
+    let config = aws_config::load_config().await;
+    // does not consume config so that config can be used to construct other service clients.
+    let client = aws_sdk_dynamodb::Client::new(&config);
+    let tables = client.list_tables().await?;
+}
+```
+
+## Proposed Design
+
+Achieving this design consists of three high level changes:
+
+1. Add a `SharedConfig` struct to `aws-types`. This contains a config, but with no logic to be loaded from the
+   environment in any way.
+2. Add an `aws-config` crate. `aws-config` contains the logic to load configuration from the environment. No generated
+   service clients will depend on `aws-config`. This is critical to avoid circular dependencies and to
+   allow `aws-config` to depend on other AWS services. `aws-config` will contain a AWS-generic configuration
+   representation + default resolver chains.
```

Comment:
Is there a risk that the naming `aws-config` would conflict with the _service_ bindings for AWS config?

(edit: reads more, sees that the service bindings are called `aws-sdk-config`.)